File: sdk/dart/test/replay_test.dart
import 'dart:async';

import 'package:ascp_sdk_dart/client.dart';
import 'package:ascp_sdk_dart/models.dart';
import 'package:ascp_sdk_dart/replay.dart';
import 'package:ascp_sdk_dart/transport.dart';
import 'package:test/test.dart';

void main() {
  group('replay helper layer', () {
    test(
      'replays historical events from from_seq and preserves replay and cursor boundaries',
      () async {
        final transport = _ReplayTransport();
        final client = AscpClient(transport: transport);

        transport.responses['sessions.subscribe'] =
            const AscpTransportSuccessResponse(
              jsonrpc: '2.0',
              id: 'req_sub_seq_1',
              result: <String, Object?>{
                'subscription_id': 'sub_replay_2',
                'session_id': 'sess_replay_1',
                'snapshot_emitted': false,
              },
            );
        transport.responses['sessions.unsubscribe'] =
            const AscpTransportSuccessResponse(
              jsonrpc: '2.0',
              id: 'req_unsub_seq_1',
              result: <String, Object?>{
                'subscription_id': 'sub_replay_2',
                'unsubscribed': true,
              },
            );

        final request = replayFromSeq(
          sessionId: 'sess_replay_1',
          fromSeq: 104,
          includeSnapshot: false,
        );

        expect(buildReplaySubscribeParams(request), <String, Object?>{
          'session_id': 'sess_replay_1',
          'from_seq': 104,
          'include_snapshot': false,
        });

        final subscription = await subscribeWithReplay(
          client: client,
          request: request,
        );

        transport.emitAll(<AscpEventEnvelope>[
          _event(
            id: 'evt_hist_104',
            type: 'message.user',
            sessionId: 'sess_replay_1',
            seq: 104,
          ),
          _event(
            id: 'evt_hist_105',
            type: 'message.assistant.delta',
            sessionId: 'sess_replay_1',
            seq: 105,
          ),
          _event(
            id: 'evt_hist_106',
            type: 'message.assistant.completed',
            sessionId: 'sess_replay_1',
            seq: 106,
          ),
          _event(
            id: 'evt_replayed_106',
            type: 'sync.replayed',
            sessionId: 'sess_replay_1',
            payload: const <String, Object?>{
              'from_seq': 104,
              'to_seq': 106,
              'event_count': 3,
            },
          ),
          _event(
            id: 'evt_live_107',
            type: 'message.assistant.delta',
            sessionId: 'sess_replay_1',
            seq: 107,
          ),
          _event(
            id: 'evt_cursor_107',
            type: 'sync.cursor_advanced',
            sessionId: 'sess_replay_1',
            payload: const <String, Object?>{'cursor': 'seq:107'},
          ),
        ]);

        final items = await subscription.stream.take(6).toList();

        expect(items.map((item) => item.kind), <String>[
          'replay_event',
          'replay_event',
          'replay_event',
          'replay_complete',
          'live_event',
          'cursor_advanced',
        ]);
        expect(subscription.subscribeResult.subscriptionId, 'sub_replay_2');
        expect(subscription.cursor, 'seq:107');
        expect(subscription.lastReplayed?.payload.eventCount, 3);
        expect(items.last.cursor, 'seq:107');
        expect(items.last.streamPhase, 'live');

        await subscription.close();

        expect(transport.recordedMethods.last, 'sessions.unsubscribe');
      },
    );

    test(
      'preserves snapshot and opaque cursor extensions without overwriting core subscribe params',
      () async {
        final request = replayWithOpaqueCursor(
          sessionId: 'sess_replay_1',
          includeSnapshot: true,
          extensionFields: const <String, Object?>{
            'cursor_token': 'opaque-123',
          },
        );

        expect(buildReplaySubscribeParams(request), <String, Object?>{
          'session_id': 'sess_replay_1',
          'include_snapshot': true,
          'cursor_token': 'opaque-123',
        });

        expect(
          () => replayWithOpaqueCursor(
            sessionId: 'sess_replay_1',
            extensionFields: const <String, Object?>{'from_seq': 104},
          ),
          throwsA(isA<AscpReplayConfigurationException>()),
        );
      },
    );
  });
}

AscpEventEnvelope _event({
  required String id,
  required String type,
  required String sessionId,
  int? seq,
  Map<String, Object?> payload = const <String, Object?>{},
}) {
  return AscpEventEnvelope(
    id: id,
    type: type,
    ts: '2026-04-22T08:16:00Z',
    sessionId: sessionId,
    seq: seq,
    payload: payload,
  );
}

final class _ReplayTransport implements AscpTransport {
  @override
  String get kind => 'stub';

  final StreamController<AscpEventEnvelope> _events =
      StreamController<AscpEventEnvelope>.broadcast();
  final Map<String, AscpTransportResponse> responses =
      <String, AscpTransportResponse>{};
  final List<String> recordedMethods = <String>[];

  @override
  Stream<AscpEventEnvelope> get events => _events.stream;

  @override
  Future<void> close() async {
    await _events.close();
  }

  @override
  Future<void> connect() async {}

  void emitAll(List<AscpEventEnvelope> events) {
    for (final event in events) {
      _events.add(event);
    }
  }

  @override
  Future<AscpTransportResponse> request(
    String method, {
    Map<String, Object?>? params,
    AscpTransportRequestOptions? options,
  }) async {
    recordedMethods.add(method);
    final response = responses[method];
    if (response == null) {
      throw StateError('No response configured for $method.');
    }
    return response;
  }
}

File: sdk/typescript/test/replay.test.ts
import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import type { AscpClientRequestOptions } from "../src/client/index.js";
import { AscpClient } from "../src/client/index.js";
import type { CoreMethodName, SessionsSubscribeResult } from "../src/methods/index.js";
import type { EventEnvelope, FlexibleObject } from "../src/models/index.js";
import { AsyncQueue } from "../src/transport/async-queue.js";
import type {
  AscpTransport,
  AscpTransportSubscription,
  CoreMethodParamsMap
} from "../src/transport/index.js";
import type { CoreMethodResponse } from "../src/validation/index.js";

const TEST_DIR = dirname(fileURLToPath(import.meta.url));
const SDK_DIR = resolve(TEST_DIR, "..");
const REPO_DIR = resolve(SDK_DIR, "..");
const REPLAY_FIXTURE_DIR = resolve(REPO_DIR, "../conformance/fixtures/replay");

type ReplayFixture = {
  request: {
    params: {
      session_id: string;
      from_seq?: number;
      from_event_id?: string;
      include_snapshot?: boolean;
    };
  };
  request_extension?: {
    token: string;
  };
  response: {
    result: SessionsSubscribeResult;
  };
  expectations: Record<string, unknown>;
  stream: EventEnvelope[];
};

type ReplayRequestLike =
  | {
      kind: "from_seq";
      params: {
        session_id: string;
        from_seq: number;
        include_snapshot?: boolean;
      };
    }
  | {
      kind: "from_event_id";
      params: {
        session_id: string;
        from_event_id: string;
        include_snapshot?: boolean;
      };
    }
  | {
      kind: "opaque_cursor";
      params: {
        session_id: string;
        include_snapshot?: boolean;
      };
      params_extension: FlexibleObject;
    };

type ReplayStreamItemLike = {
  kind: string;
  event: EventEnvelope;
  cursor?: string;
  stream_phase?: string;
};

type ReplaySubscriptionLike = AsyncIterable<ReplayStreamItemLike> & {
  close(): Promise<void>;
  cursor: string | null;
  last_replayed: EventEnvelope | null;
  last_snapshot: EventEnvelope | null;
  subscribe_result: SessionsSubscribeResult;
};

type ReplayModuleLike = {
  buildReplaySubscribeParams: (request: ReplayRequestLike) => FlexibleObject;
  replayAfterEventId: (params: {
    session_id: string;
    from_event_id: string;
    include_snapshot?: boolean;
  }) => ReplayRequestLike;
  replayFromSeq: (params: {
    session_id: string;
    from_seq: number;
    include_snapshot?: boolean;
  }) => ReplayRequestLike;
  replayWithOpaqueCursor: (options: {
    params: {
      session_id: string;
      include_snapshot?: boolean;
    };
    params_extension: FlexibleObject;
  }) => ReplayRequestLike;
  subscribeWithReplay: (
    client: AscpClient,
    request: ReplayRequestLike,
    options?: AscpClientRequestOptions
  ) => Promise<ReplaySubscriptionLike>;
};

class QueueSubscription implements AscpTransportSubscription {
  constructor(private readonly queue: AsyncQueue<EventEnvelope>) {}

  async close(): Promise<void> {
    this.queue.close();
  }

  [Symbol.asyncIterator](): AsyncIterator<EventEnvelope> {
    return this.queue[Symbol.asyncIterator]();
  }
}

class ReplayStubTransport implements AscpTransport {
  readonly kind = "stub";
  readonly requests: Array<{
    method: CoreMethodName;
    options?: AscpClientRequestOptions;
    params?: CoreMethodParamsMap[CoreMethodName] | FlexibleObject;
  }> = [];

  private readonly queue = new AsyncQueue<EventEnvelope>();
  private readonly responses = new Map<CoreMethodName, CoreMethodResponse<CoreMethodName>>();

  async connect(): Promise<void> {}

  async close(): Promise<void> {
    this.queue.close();
  }

  setResponse(
    method: CoreMethodName,
    response: CoreMethodResponse<CoreMethodName>
  ): void {
    this.responses.set(method, response);
  }

  emitAll(events: EventEnvelope[]): void {
    for (const event of events) {
      this.queue.push(event);
    }

    this.queue.close();
  }

  async request<TMethod extends CoreMethodName>(
    method: TMethod,
    params?: CoreMethodParamsMap[TMethod],
    options?: AscpClientRequestOptions
  ): Promise<CoreMethodResponse<TMethod>> {
    this.requests.push({
      method,
      ...(params !== undefined ? { params } : {}),
      ...(options !== undefined ? { options } : {})
    });

    const response = this.responses.get(method);

    if (response === undefined) {
      throw new Error(`No response configured for ${method}.`);
    }

    return response as CoreMethodResponse<TMethod>;
  }

  subscribe(): AscpTransportSubscription {
    return new QueueSubscription(this.queue);
  }
}

function loadReplayFixture(name: string): ReplayFixture {
  return JSON.parse(
    readFileSync(resolve(REPLAY_FIXTURE_DIR, name), "utf8")
  ) as ReplayFixture;
}

async function loadReplayModule(): Promise<ReplayModuleLike> {
  const module = (await import("../src/replay/index.js")) as Partial<ReplayModuleLike>;

  expect(module.replayFromSeq).toBeTypeOf("function");
  expect(module.replayAfterEventId).toBeTypeOf("function");
  expect(module.replayWithOpaqueCursor).toBeTypeOf("function");
  expect(module.buildReplaySubscribeParams).toBeTypeOf("function");
  expect(module.subscribeWithReplay).toBeTypeOf("function");

  return module as ReplayModuleLike;
}

async function collectItems(
  subscription: ReplaySubscriptionLike,
  count: number
): Promise<ReplayStreamItemLike[]> {
  const items: ReplayStreamItemLike[] = [];

  for await (const item of subscription) {
    items.push(item);

    if (items.length === count) {
      break;
    }
  }

  return items;
}

describe("replay helper layer", () => {
  it("replays historical events from from_seq and preserves replay and cursor boundaries", async () => {
    const replay = await loadReplayModule();
    const fixture = loadReplayFixture("subscribe-from-seq.json");
    const transport = new ReplayStubTransport();
    const client = new AscpClient({ transport });

    transport.setResponse("sessions.subscribe", {
      jsonrpc: "2.0",
      id: "req_sub_seq_1",
      result: fixture.response.result
    });
    transport.setResponse("sessions.unsubscribe", {
      jsonrpc: "2.0",
      id: "req_unsub_seq_1",
      result: {
        subscription_id: fixture.response.result.subscription_id,
        unsubscribed: true
      }
    });

    const request = replay.replayFromSeq({
      session_id: fixture.request.params.session_id,
      from_seq: fixture.request.params.from_seq as number,
      include_snapshot: fixture.request.params.include_snapshot
    });

    expect(replay.buildReplaySubscribeParams(request)).toEqual(fixture.request.params);

    const subscription = await replay.subscribeWithReplay(client, request);
    transport.emitAll(fixture.stream);
    const items = await collectItems(subscription, fixture.stream.length);

    expect(subscription.subscribe_result).toEqual(fixture.response.result);
    expect(transport.requests[0]).toMatchObject({
      method: "sessions.subscribe",
      params: fixture.request.params
    });
    expect(items.map((item) => item.kind)).toEqual([
      "replay_event",
      "replay_event",
      "replay_event",
      "replay_complete",
      "live_event",
      "cursor_advanced"
    ]);
    expect(items[0]?.event.id).toBe("evt_hist_104");
    expect(items[3]).toMatchObject({
      kind: "replay_complete",
      event: {
        payload: {
          from_seq: fixture.expectations.from_seq,
          to_seq: fixture.expectations.to_seq,
          event_count: fixture.expectations.replayed_event_count
        }
      }
    });
    expect(items[5]).toMatchObject({
      kind: "cursor_advanced",
      cursor: fixture.expectations.expected_cursor,
      stream_phase: "live"
    });
    expect(subscription.cursor).toBe(fixture.expectations.expected_cursor);
    expect(subscription.last_replayed?.type).toBe("sync.replayed");

    await subscription.close();

    expect(transport.requests.at(-1)).toMatchObject({
      method: "sessions.unsubscribe",
      params: {
        subscription_id: fixture.response.result.subscription_id
      }
    });
  });

  it("treats from_event_id as an exclusive anchor and switches to live only after sync.replayed", async () => {
    const replay = await loadReplayModule();
    const fixture = loadReplayFixture("subscribe-from-event-id.json");
    const transport = new ReplayStubTransport();
    const client = new AscpClient({ transport });

    transport.setResponse("sessions.subscribe", {
      jsonrpc: "2.0",
      id: "req_sub_event_id_1",
      result: fixture.response.result
    });
    transport.setResponse("sessions.unsubscribe", {
      jsonrpc: "2.0",
      id: "req_unsub_event_id_1",
      result: {
        subscription_id: fixture.response.result.subscription_id,
        unsubscribed: true
      }
    });

    const request = replay.replayAfterEventId({
      session_id: fixture.request.params.session_id,
      from_event_id: fixture.request.params.from_event_id as string,
      include_snapshot: fixture.request.params.include_snapshot
    });

    expect(replay.buildReplaySubscribeParams(request)).toEqual(fixture.request.params);

    const subscription = await replay.subscribeWithReplay(client, request);
    transport.emitAll(fixture.stream);
    const items = await collectItems(subscription, fixture.stream.length);

    expect(items.map((item) => item.kind)).toEqual([
      "replay_event",
      "replay_event",
      "replay_complete",
      "live_event"
    ]);
    expect(items[0]?.event.id).toBe(fixture.expectations.first_replayed_event_id);
    expect(items[2]).toMatchObject({
      kind: "replay_complete",
      event: {
        payload: {
          from_seq: fixture.expectations.from_seq,
          to_seq: fixture.expectations.to_seq,
          event_count: fixture.expectations.replayed_event_count
        }
      }
    });

    await subscription.close();
  });

  it("keeps snapshot events distinct from replayed history and live continuation", async () => {
    const replay = await loadReplayModule();
    const fixture = loadReplayFixture("subscribe-with-snapshot.json");
    const transport = new ReplayStubTransport();
    const client = new AscpClient({ transport });

    transport.setResponse("sessions.subscribe", {
      jsonrpc: "2.0",
      id: "req_sub_snapshot_1",
      result: fixture.response.result
    });
    transport.setResponse("sessions.unsubscribe", {
      jsonrpc: "2.0",
      id: "req_unsub_snapshot_1",
      result: {
        subscription_id: fixture.response.result.subscription_id,
        unsubscribed: true
      }
    });

    const request = replay.replayFromSeq({
      session_id: fixture.request.params.session_id,
      from_seq: fixture.request.params.from_seq as number,
      include_snapshot: fixture.request.params.include_snapshot
    });

    const subscription = await replay.subscribeWithReplay(client, request);
    transport.emitAll(fixture.stream);
    const items = await collectItems(subscription, fixture.stream.length);

    expect(items.map((item) => item.kind)).toEqual([
      "snapshot",
      "replay_event",
      "replay_event",
      "replay_complete",
      "live_event",
      "cursor_advanced"
    ]);
    expect(subscription.last_snapshot?.type).toBe("sync.snapshot");
    expect(items[0]?.event.id).toBe("evt_snapshot_curr_210");
    expect(items[1]?.event.seq).toBe(fixture.expectations.from_seq);
    expect(items[4]?.event.seq).toBe(207);
    expect(subscription.cursor).toBe(fixture.expectations.expected_cursor);

    await subscription.close();
  });

  it("passes opaque cursor extensions through without overwriting frozen core params", async () => {
    const replay = await loadReplayModule();
    const fixture = loadReplayFixture("subscribe-with-opaque-cursor.json");
    const transport = new ReplayStubTransport();
    const client = new AscpClient({ transport });

    transport.setResponse("sessions.subscribe", {
      jsonrpc: "2.0",
      id: "req_sub_cursor_1",
      result: fixture.response.result
    });
    transport.setResponse("sessions.unsubscribe", {
      jsonrpc: "2.0",
      id: "req_unsub_cursor_1",
      result: {
        subscription_id: fixture.response.result.subscription_id,
        unsubscribed: true
      }
    });

    const request = replay.replayWithOpaqueCursor({
      params: fixture.request.params,
      params_extension: {
        cursor: fixture.request_extension?.token
      }
    });

    expect(replay.buildReplaySubscribeParams(request)).toEqual({
      ...fixture.request.params,
      cursor: fixture.request_extension?.token
    });

    const subscription = await replay.subscribeWithReplay(client, request);
    transport.emitAll(fixture.stream);
    const items = await collectItems(subscription, fixture.stream.length);

    expect(transport.requests[0]).toMatchObject({
      method: "sessions.subscribe",
      params: {
        ...fixture.request.params,
        cursor: fixture.request_extension?.token
      }
    });
    expect(items.at(-1)).toMatchObject({
      kind: "cursor_advanced",
      cursor: fixture.expectations.expected_cursor
    });
    expect(subscription.cursor).toBe(fixture.expectations.expected_cursor);

    await subscription.close();
  });

  it("rejects opaque cursor extensions that try to override core subscribe params", async () => {
    const replay = await loadReplayModule();

    const request = replay.replayWithOpaqueCursor({
      params: {
        session_id: "sess_replay_1",
        include_snapshot: false
      },
      params_extension: {
        session_id: "sess_override"
      }
    });

    expect(() => replay.buildReplaySubscribeParams(request)).toThrowError(
      /must not overwrite core subscribe params/i
    );
  });
});
